bip-0174.mediawiki

```diff
@@ -358,6 +364,23 @@ for input,i in enumerate(psbt.inputs):
         assert False
 </pre>
 
+====Change Detection====
+
+Signers may wish to display the inputs and outputs to users for extra verification.
+In such displays, signers may wish to identify which outputs are change outputs in order to omit them to avoid additional user confusion.
+In order to detect change, a signer can use the BIP 32 derivation paths provided in inputs and outputs as well as the extended public keys provided globally.
+
+For a single key output, a signer can observe whether the master fingerprint for the public key for that output belongs to itself.
+If it does, it can then derive the public key at the specified derivation path and check whether that key is the one present in that output.
+
+For outputs involving multiple keys, a signer can first examine the inputs that it is signing. It should determine the general pattern of the script.
+For example, it can observe the redeemScript or witnessScript and determine whether it is multisig.
+If it is multisig, it should determine how many keys there are, in what position they are, and how many signers are needed in order for the input to be valid.
```

Comment:
Make sure the keys are ordered via BIP-67 else there will be problems.

Reply:
No. Enforcing multisig rules is out of scope for PSBT. This section is just to clarify how xpubs can be used for change detection. It isn't a description of what must be done.

Reply:
Then remove `in what position they are` because it assumes there is no ordering...

Reply:
If you specify `in what position they are` it means that if there is an input with `2 PubA PubB MULTISIG`, then a change with `2 PubB' PubA' MULTISIG` should not be considered a change.

Reply:
Position matters in some cases. We could add an example:
If `A`, `B` and `C` are xpubs, `ai`, `bi`, `ci` are keys derived from them and input uses a spending script with condition `(a1 or b1) and c1` then output with condition `(a2 or b2) and c2` or `(b2 or a2) and c2` can be considered as change, but `(a2 or c2) and b2` should not.

Reply:
I've made it clearer that those things are examples of things that signers can check, not things that they must check.

Reply:
@stepansnigirev no, if you follow BIP67, then `(a2 or c2) and b2` SHOULD be considered as change and not `(b2 or a2) and c2`.

@achow101 please remove the `in what position they are` this will really cause problem in the future. Either you force implementer to check BIP67, or you don't imply they can check the order.

If you say they can check the order, and that they do, we will have bunch of incompatibilities accross HW and wallet interfaces.

Reply:
@NicolasDorier The BIP is not saying anything about what signers must check. That whole section is about constructing the policy represented by a script, i.e. the descriptor. Public key order is something that can be part of such constructions. At no point does it say that signers must check that or that public key order must be part of the policy.

Reply:
> At no point does it say that signers must check that or that public key order must be part of the policy.

But that's my problem! it says that the public key order must be part of the policy:

> If it is multisig, it should determine how many keys there are, **in what position they are**

Reply:
How about you read and review the current text instead of an outdated one? It no longer says that!

Reply:
Indeed! The updated one seems OK to me.